File: DataMaker.Sdk.Js/src/validate.js
'use strict';

const { normalizeKind, isInputKind, coerceValue, NON_INPUT_KINDS } = require('./fieldKinds');

// Validate + coerce a caller's values object against a form's field schema.
// Returns { values, issues }:
//   values — the coerced dict keyed by field name, ready to seal. Only known,
//            non-empty, input-kind fields land here.
//   issues — [{ field, kind, message }]; empty means the input is clean.
//
// Rules: unknown keys are rejected (typo protection), values for read-only
// kinds (calc/heading/signature) are rejected, required fields must be present
// and non-empty, and every value is coerced to its kind's wire shape.
function validateValues(fields, input, opts = {}) {
  const allowUnknown = opts.allowUnknown === true;
  const issues = [];
  const values = {};

  const byKey = new Map();
  for (const f of fields) byKey.set(f.key, f);

  for (const [key, raw] of Object.entries(input || {})) {
    const field = byKey.get(key);
    if (!field) {
      if (!allowUnknown) {
        issues.push({ field: key, kind: null, message: 'unknown field — not in the form schema' });
      }
      continue;
    }
    if (!isInputKind(field.kind)) {
      issues.push({
        field: key,
        kind: normalizeKind(field.kind),
        message: `field is read-only (${normalizeKind(field.kind)}) and cannot be submitted`,
      });
      continue;
    }
    if (isEmpty(raw)) continue; // handled by the required pass below

    const { value, error } = coerceValue(field.kind, raw, field);
    if (error) {
      issues.push({ field: key, kind: normalizeKind(field.kind), message: error });
      continue;
    }
    values[key] = value;
  }

  for (const field of fields) {
    if (!field.required) continue;
    if (NON_INPUT_KINDS.has(normalizeKind(field.kind))) continue;
    if (!(field.key in values)) {
      issues.push({
        field: field.key,
        kind: normalizeKind(field.kind),
        message: 'required field is missing',
      });
    }
  }

  return { values, issues };
}

function isEmpty(v) {
  if (v === undefined || v === null) return true;
  if (typeof v === 'string') return v.trim() === '';
  if (Array.isArray(v)) return v.length === 0;
  return false;
}

module.exports = { validateValues };

File: DataMaker.Sdk.Py/src/datamaker/validate.py
"""Validate + coerce a caller's values dict against a form's field schema."""

from __future__ import annotations

from typing import Any, Dict, List, Tuple

from .fields import normalize_kind, is_input_kind, coerce_value, NON_INPUT_KINDS


def _is_empty(v: Any) -> bool:
    if v is None:
        return True
    if isinstance(v, str):
        return v.strip() == ""
    if isinstance(v, (list, tuple)):
        return len(v) == 0
    return False


def validate_values(
    fields: List[dict], input_values: Dict[str, Any], allow_unknown: bool = False
) -> Tuple[Dict[str, Any], List[dict]]:
    """Returns ``(values, issues)``. ``values`` holds only known, non-empty,
    input-kind fields coerced to their wire shape. ``issues`` is empty when the
    input is clean."""
    issues: List[dict] = []
    values: Dict[str, Any] = {}
    by_key = {f["key"]: f for f in fields}

    for key, raw in (input_values or {}).items():
        field = by_key.get(key)
        if field is None:
            if not allow_unknown:
                issues.append({"field": key, "kind": None, "message": "unknown field — not in the form schema"})
            continue
        if not is_input_kind(field["kind"]):
            nk = normalize_kind(field["kind"])
            issues.append({"field": key, "kind": nk, "message": f"field is read-only ({nk}) and cannot be submitted"})
            continue
        if _is_empty(raw):
            continue
        value, error = coerce_value(field["kind"], raw, field)
        if error:
            issues.append({"field": key, "kind": normalize_kind(field["kind"]), "message": error})
            continue
        values[key] = value

    for field in fields:
        if not field.get("required"):
            continue
        if normalize_kind(field["kind"]) in NON_INPUT_KINDS:
            continue
        if field["key"] not in values:
            issues.append({"field": field["key"], "kind": normalize_kind(field["kind"]), "message": "required field is missing"})

    return values, issues
